Shared/Confirm-ExchangeShell.ps1

```diff
@@ -1,111 +1,95 @@
 ﻿# Copyright (c) Microsoft Corporation.
 # Licensed under the MIT License.
 
+. $PSScriptRoot\Invoke-CatchActionError.ps1
+. $PSScriptRoot\Invoke-CatchActionErrorLoop.ps1
+
 Function Confirm-ExchangeShell {
     [CmdletBinding()]
     param(
         [Parameter(Mandatory = $false)][bool]$LoadExchangeShell = $true,
         [Parameter(Mandatory = $false)][scriptblock]$CatchActionFunction
     )
 
-    Function Invoke-CatchActionErrorLoop {
-        param(
-            [int]$CurrentErrors
-        )
-
-        if ($null -ne $CatchActionFunction -and
-            $Error.Count -ne $CurrentErrors) {
-            $i = 0
-            while ($i -lt ($Error.Count - $currentErrors)) {
-                & $CatchActionFunction $Error[$i]
-                $i++
-            }
-        }
+    begin {
+        $passed = $false
+        $edgeTransportKey = 'HKLM:\SOFTWARE\Microsoft\ExchangeServer\v15\EdgeTransportRole'
+        $setupKey = 'HKLM:\SOFTWARE\Microsoft\ExchangeServer\v15\Setup'
+        Write-Verbose "Calling: $($MyInvocation.MyCommand)"
+        Write-Verbose "Passed: LoadExchangeShell: $LoadExchangeShell"
     }
-
-    $passed = $false
-    $edgeTransportKey = 'HKLM:\SOFTWARE\Microsoft\ExchangeServer\v15\EdgeTransportRole'
-    $setupKey = 'HKLM:\SOFTWARE\Microsoft\ExchangeServer\v15\Setup'
-    Write-VerboseWriter("Calling: Confirm-ExchangeShell")
-    Write-VerboseWriter("Passed: [bool]LoadExchangeShell: {0}" -f $LoadExchangeShell)
-
-    try {
-        $currentErrors = $Error.Count
-        Get-ExchangeServer -ErrorAction Stop | Out-Null
-        Write-VerboseWriter("Exchange PowerShell Module already loaded.")
-        $passed = $true
-        Invoke-CatchActionErrorLoop -CurrentErrors $currentErrors
-    } catch {
-        Write-VerboseWriter("Failed to run Get-ExchangeServer")
-
-        if ($null -ne $CatchActionFunction) {
-            & $CatchActionFunction
-        }
-
-        if (!$LoadExchangeShell) {
-            return [PSCustomObject]@{
-                ShellLoaded = $false
+    process {
+        try {
+            $currentErrors = $Error.Count
+            Get-ExchangeServer -ErrorAction Stop | Out-Null
```

Comment:
We should add -ResultSize 1 or use a cheaper/faster cmdlet to test, like Get-OrganizationConfig or Get-ADServerSettings.

Reply:
Will Add `-ResultsSize 1` as we have run into issues with `Get-OrganizationConfig` not working due to RBAC for some users.

Reply:
@bill-long @lusassl-msft Let me know what you think. 

Actually, `Get-ExchangeServer` doesn't have an option for `-ResultSize`.

https://docs.microsoft.com/en-us/powershell/module/exchange/get-exchangeserver?view=exchange-ps

Because there has been a few cases where `Get-OrganizationConfig` isn't able to run by some people, don't want this function to fail because of that. So i think that should be out of the question. 

`Get-ExchangeServer` is the most common command to get information about the server that is typically run in scripts, or at least the main ones that i have written. We could do `Get-ADServerSettings` which does give us more RBAC options to help confirm that we are in an Exchange Management Shell, or we can include a parameter to pass an `-Identity` to `Get-ExchangeServer`


![image](https://user-images.githubusercontent.com/22776718/127687023-4fb99f9d-7eff-423c-a21e-f17a457e1444.png)

Reply:
If we're just trying to confirm the command is available, we test for:

`($null -ne (Get-Command Get-ExchangeServer -ErrorAction SilentlyContinue))`

If True, something has loaded a command called Get-ExchangeServer.

Reply:
Yes, we can do that. However, the secondary purpose of this functionality of the function is to determine if we are in a stale or idle session. If we are, when we run `Get-ExchangeServer` it ends up reconnecting the connection. It handles the errors that might occur and if we failed to re-connect it tries to create a new session similar as how we connect when launching the shortcut of EMS. That is why we can't just test to see if the cmdlet is there.

Reply:
We could explicitly check for an open session:

![image](https://user-images.githubusercontent.com/4518572/127693234-ec373b33-c7ff-48ef-8b56-ac4dd4857c14.png)
...
![image](https://user-images.githubusercontent.com/4518572/127693348-57e07049-f814-4617-9c0c-24b50c53708b.png)

Though if we have to run Get-ExchangeServer (or some other random cmdlet) to re-establish it anyway, I guess there's not much benefit, and we should just stick with the random cmdlet call.

Reply:
It does reconnect most of the time just fine. However, it also throws errors while reconnecting. Within the Health Checker, there is built in logic to determine if there were errors that weren't handled. If we determine there was an error we didn't handle, at the end of the script we provide to email with the debug log to determine the cause. 

The other option like I provided was to pass a server name too. That will also limit the call and make it faster for large environments. The other option might be to do `Get-ADServerSettings`.

Reply:
Yeah, passing a single server name would be as good as -ResultSize 1.

Reply:
Fixed in 02dd28d07e442085124aff8d8009d802572c3f7a